feat: Support for MariaDB 10.6

gavindsouza opened the pull request:
Okay, maybe the title is a little incorrect; The right way to describe it would be _"`COMPRESSED` is deprecated in MariaDB 10.6, it will be removed in 10.7. Let's get ready for that"_.

With MariaDB 10.6 tables using ROW_FORMAT=COMPRESSED are read-only so if you try to install Frappe with MariaDB 10.6 you get the error **InnoDB refuses to write tables with ROW_FORMAT=COMPRESSED or KEY_BLOCK_SIZE**. The setting ROW_FORMAT=COMPRESSED is deprecated in MariaDB 10.6. This setting is scheduled to be removed in MariaDB 10.7.

Now, if you don't want to make any changes to your tables right away, you can still work with MariaDB 10.6 by telling it to simply *not* treat tables with `COMPRESSED` row_format as read only. **Workaround:** Add innodb_read_only_compressed=OFF to the MariaDB configuration file(Under [mysqld] group) and restart MariaDB, or run SET GLOBAL innodb_read_only_compressed=OFF. (ref: [innodb_read_only_compressed](https://mariadb.com/docs/reference/mdb/system-variables/innodb_read_only_compressed/))

---

### What this PR brings into Frappe?

- Added command to transform tables: Ability to change table's `ENGINE` and `ROW_FORMAT` via the **bench transform-table** command.

<img width="722" alt="Screenshot 2021-09-02 at 4 07 51 PM" src="https://user-images.githubusercontent.com/36654812/131829451-fff5a067-fe00-4f6b-99b0-9a6fa8c132e1.png">

- Added common command to access database console: **bench db-console** instead of bench mariadb or bench postgres...why remember what dbms you're running the site on??? 🙄 
- Update restore command: Replace sql text stream from "ROW_FORMAT=COMPRESSED" to "ROW_FORMAT=DYNAMIC"
- Updates CI MariaDB version (tests restore command and default row_format DYNAMIC)
- New `row_format` default will be DYNAMIC. This can be changed later if you wish using the transform-database command, or any other means.

---

### Other relevant info

From @sagarvora's comment in the thread:
> If we want compressed-like efficiency for data storage (don't know the performance / storage space implications), MariaDB provides an [alternate way](https://mariadb.com/kb/en/innodb-page-compression/) to configure it.

Benchmarks run to compare COMPRESSED vs DYNAMIC row_format changes, from @sunnyakaxd's comment below: https://github.com/frappe/frappe/pull/13954#issuecomment-901918179


---

Fixes https://github.com/frappe/frappe/issues/13810
Docs: https://github.com/frappe/frappe_docs/pull/199

gavindsouza committed:
feat: Add util to convert compressed tables to DYNAMIC

NOTE: This shouldn't be considered as a finished command. Only run this
if you understand what you're doing

gavindsouza committed:
chore: Use DYNAMIC row_format instead of deprecated COMPRESSED

refs:
* MariaDB 10.6 gets rid of the COMPRESSED row_format
* https://dev.mysql.com/worklog/task/?id=8307
* https://stackoverflow.com/questions/24321896/mysql-row-format-compressed-vs-dynamic

sagarvora commented on frappe/commands/utils.py:
```diff
@@ -512,6 +512,29 @@ def console(context):
+@click.command('convert-database')
+@pass_context
+def convert_database(context):
+	"convert row_formats to DNAMIC from older formats -- innodb mariadb v10.6.3"
```
```suggestion
	"convert row_formats to DYNAMIC from older formats -- innodb mariadb v10.6.3"
```

sagarvora commented:
If we want compressed-like efficiency for data storage (don't know the performance / storage space implications), MariaDB provides an [alternate way](https://mariadb.com/kb/en/innodb-page-compression/) to configure it.

sunnyakaxd commented:
I wonder if this helps, we tried out profiling dynamic row format a bit
```
{
 "test_innodb_Compressed": {
  "data_mb:Float:100": 393.76,
  "index_mb:Float:100": 0.0,
  "mb:Float:100": 393.76,
  "throughput": {
   "bulk_read_time": 2278.847945,
   "per_insert_ms": 31.831568237759775,
   "reads_per_ms": 5.802483163254668
  }
 },
 "test_innodb_Dynamic": {
  "data_mb:Float:100": 784.52,
  "index_mb:Float:100": 0.0,
  "mb:Float:100": 784.52,
  "throughput": {
   "bulk_read_time": 762.654343,
   "per_insert_ms": 16.556692576960774,
   "reads_per_ms": 6.543929148988272
  }
 }
}
```

Dynamic row format makes bulk reads and inserts a lot faster, though takes almost twice as space.

ankush commented on frappe/commands/utils.py:
```diff
@@ -512,6 +512,29 @@ def console(context):
+	for table in tables:
+		frappe.db.sql(f"ALTER TABLE `{table}` ROW_FORMAT=DYNAMIC")
+
+	frappe.db.commit()
```
```suggestion
```

Not required for ddl? 

---

Also `db == mariadb` should probably be asserted before running this.

gavindsouza committed:
feat: Handle site restores to MariaDB 10.6

gavindsouza committed:
feat(cli): New command 'db-console'

Don't remember what db console you have to open for what site!

How to use: `bench --site framework.io db-console`

gavindsouza committed:
feat: Transform tables for given site

* Ability to update ROW_FORMAT for mentioned tables in --table option
* Failfast to stop on first error occured
* Show progressbar for overall conversion progress (remaining tales)

gavindsouza committed:
feat(minor): transform-tables - Update table engine

gavindsouza committed:
Merge branch 'develop' into mariadb-10_6

gavindsouza committed:
perf: Single query to update a table

gavindsouza committed:
fix(restore): Initialize site before decompressing file

gavindsouza commented:
I've written a patch to convert all non-dynamic row_formats with InnoDB to DYNAMIC but haven't pushed it yet. The reasons for not adding a patch for this tables conversion are:

1. You don't have to convert all your tables to `DYNAMIC` for the time being, if you're on anything older than `10.6`. If you're on `10.6`, you can still work with it by simply changing a variable globally.
2. So, this only affects you when you upgrade mariaDB locally or migrate to a newer server. You can migrate the site whenever you feel it's required (maybe now, maybe after you migrate to mariaDB 10.7) by running `bench --site site transform-database --tables all --row_format DYNAMIC`
3. A patch like this could considerably increase disk usage for large-scale deployments (or smaller ones...idk their workflows) [ref: https://github.com/frappe/frappe/pull/13954#issuecomment-901918179]. I don't want to change this implicitly during a version upgrade.
4. Some deployments may have selected different row_formats for different tables to fine-tune their sites.

hasnain2808 commented:
Tested Both Methods

gavindsouza committed:
chore: Add help for bench transform-database

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into mariadb-10_6

pm-at commented:
Can this patch be backported to v13?

gbm001 commented:
The deprecation of ROW_FORMAT=COMPRESSED is cancelled, although there are issues for Mariadb 10.6.0 to 10.6.5 (just needs an extra flag in the options to set the tables to writeable).

https://jira.mariadb.org/browse/MDEV-22367
